Add gamepad rumble support to bevy_input

johanhelsing opened the pull request:
# Objective

Provide the ability to trigger controller rumbling (force-feedback) with a cross-platform API.

## Solution

This adds the `GamepadRumbleRequest` event to `bevy_input` and adds a system in `bevy_gilrs` to read them and rumble controllers accordingly.

It's a very primitive API with a `RumbleIntensity` and `duration` in seconds that is easy to    understand and use.

I used #3868 as a starting point, updated to main, removed the low-level gilrs effect API, and moved the requests to `bevy_input`.

I intend this to hopefully be a non-controversial cross-platform starting point we can build upon to eventually support more fine-grained control (closer to the gilrs effect API)

---

## Changelog

### Added

- Gamepads can not rumble using the `GamepadRumbleRequest` event.

johanhelsing commented on crates/bevy_gilrs/src/rumble.rs:
```diff
@@ -0,0 +1,99 @@
+    let current_time = time.elapsed_seconds();
+    // Remove outdated rumble effects.
+    // `ff::Effect` uses RAII, dropping = deactivating
+    manager
```
@nicopap I removed an `is_empty` check here because I thought it wasn't needed... but I guess maybe it was there to prevent triggering change detection?

nicopap replied:
That was to avoid change detection, it's triggered as soon as a `&mut T` is created from the `NonSendMut<T>`. The `is_empty()` prevents that. I don't recall why I thought it was important though.

johanhelsing replied:
Okay, assuming it's not worth the extra clutter, then.

johanhelsing commented on crates/bevy_gilrs/src/rumble.rs:
```diff
@@ -0,0 +1,99 @@
+fn get_effect_type(intensity: RumbleIntensity) -> ff::BaseEffectType {
+    use RumbleIntensity::*;
+    match intensity {
+        Strong => ff::BaseEffectType::Strong { magnitude: 63_000 },
+        Medium => ff::BaseEffectType::Strong { magnitude: 40_000 },
+        Weak => ff::BaseEffectType::Weak { magnitude: 40_000 },
```
@alice-i-cecile raised a concern in the previous PR was that these values seemed somewhat arbitrary.

@nicopap do you remember why you chose these values specifically? Is there another less arbitrary scale we could use?

nicopap replied:
It's fully arbitrary. I picked them based on how it felts in my hand with my controller (a logitech F710).

johanhelsing replied:
I did some digging.

- The scale in gilrs goes from 0 to `u16::MAX`, because that's what [xinput does](https://learn.microsoft.com/en-us/windows/win32/api/xinput/ns-xinput-xinput_vibration?source=recommendations)
- The `Strong` and `Weak` `BaseEffectType`s map to left or right motors in the windows API. On linux, they map to `strong_magnitude` and `weak_magnitude` on the `ff_rumble` ioctl, on wasm and macos, it does nothing.
- I *think* the convention is that the left motor in most controllers is a low frequency motor, while the right is a high frequency.
- I think It might make sense to spin both motors at full speed for maximum rumble.

So it seems that on the two platforms they don't perfectly agree on which motor is which, but by convention it kind of is the same... So I'm wondering how to present a unified, and easy to use API.

I'm thinking it would at least make sense to make the scale be an `f32` from 0.0 to 1.0. Using `u16` seems a bit too implementation-specific and low level to be convenient to use, even though both windows xinput and linux ioctls use it.

For how to deal with the separate motors and speeds, I see the following options:

1. Keep it almost as is for now, but increase the Strong level to `u16::MAX`.
2. Add an `enum RumbleMotor { Strong, Weak }` and expose it in the event.

I also think it probably makes sense to make the effects additive instead of overwriting, at least by default.

If we want to provide some way to cancel running rumbles, we should rather return some kind of handle instead. Perhaps it would make sense to have a `Rumbles` resource instead of using events?

johanhelsing replied:
3. Remove intensity in this PR and just provide duration.

johanhelsing replied:
checked a few other game engines. godot also uses separate "strong" and "weak" magnitudes that go from 0 to 1: https://docs.godotengine.org/en/stable/classes/class_input.html#class-input-method-start-joy-vibration

nicopap replied:
Awesome! Thanks for taking the time to clear things up.  I think the gilrs API is very weird (as you uncovered here!) and we should strive to wrap it into something sensible.

When making a rumble API, I see two questions to answer:

1. How should the user "talk" with the rumble system?
2. How to manage the "high frequency" and "low frequency" motors?

### 1. Talking with gamepads

I see three options:

- **gamepads as entities**: Gamepad state (including rumble) could be encoded as components. This also makes it easier to interact with gamepads in general.
- **rumble resource**: the least controversial API would be to add a new resource containing a `Vec<Rumble>`, (or `HashMap<GamepadId, Rubmle>`) where `Rumble` has a single field `intensity: f32`. The `RumbleManagers` would be updated by a system that reads the rumble resource.
- **event API**:

```rust
enum RumbleRequest {
    Start { gamepad: GamepadId, duration: f32, intensity: f32 },
    Stop { gamepad: GamepadId },
}
```

### 2. Specifying motors

I think the API should either forego individual motor controls or be clear about what `Strong` and `Weak` mean. Those are very misleading names, since they don't describe what they represent. They should be clearly documented in a `GamepadRumbleMotor` enum, or fields of whatever resource/component we chose to add.

---

Personally I'd go with the resource. This way the question of "overwrite vs additive" solves itself and it's possible for the user to inspect the rumble state before modifying it. Also it's a much less controversial change than gamepads as entities, even though I think that would be the most sensible solution.

It's still technically more challenging, since you have to convert it into the gilrs representation based on durations. It's also a bit harder to come up with a good internal representation.

Also yes please `f32: [0, 1]` is so much better than a `u16`, I think this is not controversial.

johanhelsing replied:
> ## 2. Specifying motors

I focused on this first: While I also kind of felt like it would have been better to have a single strength value, the problem that then presents itself, how do we map that to the two motor values exposed in the underlying API?

Do we set both motors to the same `intensity`, or would it make more sense to scale the weak motor to max before we even start enabling the strong one? I opted to expose both values, at least for now, since that's what both other engines and the lower-level APIs seem to do.

> ## 1. Talking with gamepads

I was thinking a bit about how and when you would use it. One very common case would that the character just got hit by something, fell to the ground etc. In those cases, what you want is a fire and forget style API (rumble at this intensity for this many seconds). Also, you'd probably want them to add up, you don't want a long low intensity rumble to cancel an ongoing "important" rumble (like a character losing health).

Same goes for ui buttons, you're probably mostly interested to set a given strength and duration, then forget about it.

I'm a bit uncertain how to implement that in a resource-style API... We could perhaps do something like the following?

```rust
// bevy_input
#[derive(Resource)]
pub struct GamepadRumbles(HashMap<Gamepad, Vec<GamepadRumble>>);

impl GamepadRumbles {
    fn stop_all()
    fn stop(gamepad: Gamepad)
    fn add_rumble(...)
    fn replace_rumble()
}

struct GamepadRumble {
    deadline: f32
    instensity: GamepadRumbleIntensity,
}
```

And then do book-keeping in bevy_gilrs to sync the state?

nicopap replied:
I went with events because that was how I implemented it in my game before opening the PR. So it makes sense to keep it that way. A case can be made that both modes of control could co-exist (at least for the user to be able to read the current rumble state), but I think IMO this PR could be merged even with only the event API.

With regard to individual motor controls based on events, you could have distinct events to send for the high magnitude and the low magnitude motors. But I think we should have a similar API to that of godot you linked, by having two intensity fields in the `Start` event. `hight_magnitude_motor_intensity: f32` and `low_magnitude_motor_intensity: f32` (of course, names could/should be different)

You suggestion that rumble intensity from events should be additive is growing on me even though I was initially doubtful. It solves a usability problem fairly cleanly. But I think this needs to be documented.

johanhelsing replied:
Okay, so to sum up we keep this PR mostly as is? Then perhaps revisit adding a resource in another PR?

Will add some inline comments in the latest diff with my remaining questions to keep the discussions more narrow and easier to follow.

alice-i-cecile replied:
Agreed with leaving this mostly as is. Intensity control is a must-have IMO.

johanhelsing commented on crates/bevy_input/src/gamepad.rs:
````diff
@@ -1240,6 +1240,83 @@ const ALL_AXIS_TYPES: [GamepadAxisType; 6] = [
+/// }
+/// ```
+#[derive(Clone)]
+pub struct GamepadRumbleRequest {
````
Would it be better to make the request an enum instead of having an `additive` field? Something like this?

```rust
#[derive(Clone)]
pub enum GamepadRumbleRequest {
    /// Add a rumble to the given gamepad.
    ///
    /// To replace a rumble, Send a `Stop` event first.
    Add {
        gamepad: Gamepad,
        duration_seconds: f32,
        intensity: GamepadRumbleIntensity,
    },
    Stop {
        gamepad: Gamepad,
    }
}
```

nicopap replied:
yes I think so.

alice-i-cecile replied:
Definitely agree on the enum form.

johanhelsing commented on crates/bevy_input/src/gamepad.rs:
```diff
@@ -1240,6 +1240,83 @@ const ALL_AXIS_TYPES: [GamepadAxisType; 6] = [
 ];
 
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct GamepadRumbleIntensity {
```
What do you think, does this struct make sense, or should I just keep two fields on the event instead?

nicopap replied:
It's going to be a bit annoying to construct rumble requests

```rust
GamepadRumbleRequest {
  // ...
  intensity: GamepadRumbleIntensity { strong: 1.0, weak: 0.5 },
  // ...
}
```

But  I think it's fine as-is.

johanhelsing replied:
I put in some convenience constants, so:

```rust
intensity: GamepadRumbleIntensity::WEAK_MAX
```

STRONG_MAX, and MAX. You could also define some yourself specific to your game.

johanhelsing commented on crates/bevy_input/src/gamepad.rs:
```diff
@@ -1240,6 +1240,109 @@ const ALL_AXIS_TYPES: [GamepadAxisType; 6] = [
+#[derive(Clone)]
+pub struct GamepadRumbleRequest {
+    /// The duration in seconds of the rumble
+    pub duration_seconds: f32,
```
Would it be better to use a Duration here? What do other APIs of Bevy do?

nicopap replied:
Dunno, look at `Time` and `FixedTime` resources. Looks like in existing APIs, `Duration` is used. But I think there are some edge cases regarding wasm one needs to be aware of, not sure which. `Duration` seems the most sensible even if personally I'd always prefer a `f32`.

johanhelsing replied:
Switched to Duration

johanhelsing commented:
So, the failing Dependencies job... I'm not really sure why it fails. Is there something i need to do?

Otherwise, this pr is now ready as far as I'm concerned.

alice-i-cecile commented:
Nope: that needs another PR to address and isn't your concern. It won't block CI.

alice-i-cecile commented on crates/bevy_gilrs/src/rumble.rs:
```diff
@@ -0,0 +1,146 @@
+use crate::converter::convert_gamepad_id;
+
+struct RunningRumble {
+    deadline: f32,
```
```suggestion
    /// How long until the effect expires.
    deadline: f32,
```

alice-i-cecile commented on crates/bevy_gilrs/src/rumble.rs:
```diff
@@ -0,0 +1,146 @@
+}
+
+#[derive(Default)]
+pub(crate) struct RumblesManager {
```
Doc strings please. This is used as a non-send resource below, but it's hard to figure out the intent of this at first glance.

johanhelsing replied:
I renamed it to `RunningRumbleEffects`, which should hopefully make its purpose more clear, and added a little bit for explanation.

alice-i-cecile commented on crates/bevy_gilrs/src/rumble.rs:
```diff
@@ -0,0 +1,146 @@
+    rumbles: HashMap<GamepadId, Vec<RunningRumble>>,
+}
+
+fn to_gilrs_magnitude(ratio: f32) -> u16 {
```
Doc strings for why the conversion is done this way would be very helpful for future readers.

nicopap commented on crates/bevy_gilrs/src/lib.rs:
```diff
@@ -22,8 +24,10 @@ impl Plugin for GilrsPlugin {
                     .add_systems(PreStartup, gilrs_event_startup_system)
-                    .add_systems(PreUpdate, gilrs_event_system.before(InputSystem));
+                    .add_systems(PreUpdate, gilrs_event_system.before(InputSystem))
+                    .add_systems(PostUpdate, play_gilrs_rumble);
```
Need to make `play_girls_rumble` public, or at least create a label for it.

alice-i-cecile commented on crates/bevy_gilrs/src/rumble.rs:
```diff
@@ -0,0 +1,146 @@
+) {
+    let current_time = time.elapsed_seconds();
+    // Remove outdated rumble effects.
+    for (_gamepad, rumbles) in manager.rumbles.iter_mut() {
```
We can use .values_mut() for cleaner code here.

alice-i-cecile commented on crates/bevy_gilrs/src/rumble.rs:
```diff
@@ -0,0 +1,146 @@
+        match handle_rumble_request(&mut manager, &mut gilrs, rumble, current_time) {
+            Ok(()) => {}
+            Err(RumbleError::GilrsError(err)) => {
+                debug!(
```
```suggestion
                warn!(
```

I think these errors are similar severity, and `debug` warnings are hidden by default.

johanhelsing replied:
I left force feedback not supported as a debug, since we currently don't have a way to query for force feedback support for a controller (avoid spamming warnings the user can't solve)

nicopap commented on crates/bevy_gilrs/src/rumble.rs:
```diff
@@ -0,0 +1,146 @@
+    effect: ff::Effect,
+}
+
+enum RumbleError {
+    GamepadNotFound,
+    GilrsError(ff::Error),
+}
+impl From<ff::Error> for RumbleError {
+    fn from(err: ff::Error) -> Self {
+        RumbleError::GilrsError(err)
+    }
+}
```
Even if private, and relatively trivial, I would recommend using `thiserror` here, it let you avoid the `impl From`.

alice-i-cecile commented on crates/bevy_input/src/gamepad.rs:
```diff
@@ -1240,6 +1241,104 @@ const ALL_AXIS_TYPES: [GamepadAxisType; 6] = [
+    ///
+    /// By convention, this is usually a low-frequency motor on the left-hand
+    /// side of the gamepad, though it may vary across platforms and hardware.
+    pub strong: f32,
```
I think that these fields will be clearer out of context if renamed as `strong_motor` and `weak_motor`. Setting a low value for `strong` for example becomes quite confusing otherwise!

alice-i-cecile commented on crates/bevy_input/src/gamepad.rs:
```diff
@@ -1240,6 +1241,104 @@ const ALL_AXIS_TYPES: [GamepadAxisType; 6] = [
+    };
+
+    /// Creates a new rumble intensity with weak motor intensity set to the given value
+    pub fn weak(weak: f32) -> Self {
```
These constructors can be const fn.

johanhelsing replied:
I added clamping, which means they can't anymore :/

johanhelsing replied:
Contra: Removed the clamping again (handled in bevy_gilrs), added const.

alice-i-cecile reviewed (changes requested):
Code quality and docs feedback, but I'm happy with the core design and would like to see this ship for 0.11 :)

nicopap commented on crates/bevy_gilrs/src/rumble.rs:
```diff
@@ -0,0 +1,146 @@
+    rumble: GamepadRumbleRequest,
+    current_time: f32,
+) -> Result<(), RumbleError> {
+    let gamepad = match rumble {
+        GamepadRumbleRequest::Add { gamepad, .. } | GamepadRumbleRequest::Stop { gamepad } => {
+            gamepad
+        }
+    };
```
This would be a bit cleaner if defined as a method on `GamepadRumbleRequest`.

Also you can use the properties of infallible patterns here, the following should compile:

```rust
let GamepadRumbleRequest::Add { gamepad, .. } | GamepadRumbleRequest::Stop { gamepad } = gamepad;
```

nicopap commented on crates/bevy_gilrs/src/rumble.rs:
```diff
@@ -0,0 +1,146 @@
+            intensity,
+            ..
+        } => {
+            let deadline = current_time + duration.as_secs_f32();
```
A pitfall of the deadline pattern I used in the original PR is that they don't scale well. Here after 1 or 2 hours of gameplay, this doesn't work anymore. This would work with a `f64` though.

Using `Duration` would solve it too.

nicopap commented on crates/bevy_gilrs/src/rumble.rs:
```diff
@@ -0,0 +1,146 @@
+        } => {
+            let deadline = current_time + duration.as_secs_f32();
+
+            let mut effect_builder = ff::EffectBuilder::new();
+
+            for effect in get_base_effects(intensity) {
+                effect_builder.add_effect(effect);
+            }
+
+            let effect = effect_builder.gamepads(&[gamepad_id]).finish(gilrs)?;
+            effect.play()?;
```
This API is so... hmm, different from what we settled on. The motivation behind it is laid out here <https://gitlab.com/gilrs-project/gilrs/-/merge_requests/11>. Nothing related to this code specifically, just FYI for future people.

nicopap commented on crates/bevy_gilrs/src/rumble.rs:
```diff
@@ -0,0 +1,146 @@
+
+    // Add new effects.
+    for rumble in requests.iter().cloned() {
+        let gamepad = match rumble {
+            GamepadRumbleRequest::Add { gamepad, .. } | GamepadRumbleRequest::Stop { gamepad } => {
+                gamepad
+            }
+        };
```
As mentioned earlier, this would be better as a method on `GamepadRumbleRequest`

nicopap commented on crates/bevy_gilrs/src/rumble.rs:
```diff
@@ -0,0 +1,146 @@
+    (ratio * u16::MAX as f32) as u16
+}
+
+fn get_base_effects(
```
Could this be defined as a `GamepadRumbleIntensity` method?

johanhelsing replied:
`GamepadRumbleIntensity` is in `bevy_input`, so I think it would have to be an extension method, seems like too many hoops.

nicopap commented on crates/bevy_gilrs/src/rumble.rs:
```diff
@@ -0,0 +1,146 @@
+            gamepad
+        }
+    };
+    let (gamepad_id, _) = gilrs
+        .gamepads()
+        .find(|(pad_id, _)| convert_gamepad_id(*pad_id) == gamepad)
+        .ok_or(RumbleError::GamepadNotFound)?;
```
hmm, so what happens when you `GamepadRumbleRequest::Add` with an already running request. Does gilrs documents it? You seem to add it to the list of rumbles for the given controller.

Supposedly you documented the behavior in the `GamepadRumbleRequest` doc. You did it in knowledge of the gilrs implementation right?

johanhelsing replied:
Yes, gilrs does a saturating add of multiple effects, if i understood your question properly.

nicopap replied:
I feel like this might be good to document this. It's intent that's not obviously expressed in the code, therefore warrants a comment. I suggest it as a doc string for the `RumblesManager::rumbles` field.

nicopap commented on crates/bevy_input/src/gamepad.rs:
```diff
@@ -1240,6 +1241,104 @@ const ALL_AXIS_TYPES: [GamepadAxisType; 6] = [
+    };
+
+    /// Creates a new rumble intensity with weak motor intensity set to the given value
+    pub fn weak(weak: f32) -> Self {
+        Self { weak, strong: 0.0 }
+    }
```
What do you think of asserting that `weak` is within [0.0, 1.0]. I can see people mistakenly doing a `-1.0` or `2.0` request and being surprised to get a `u16` underflow/overflow (at least on debug builds, they would just get wild results on release builds).

alice-i-cecile replied:
Strongly in favor of adding an assert here, or even new-typing a `RumbleStrength` that does saturating operations.

mockersf replied:
or it could be clamped rather than panicking

nicopap reviewed (changes requested):
That `f32` needs to be a `Duration`, and need to be careful about under/overflow on the `f32` to `u16` conversion.

Otherwise this looks good.

nicopap commented on crates/bevy_input/src/gamepad.rs:
```diff
@@ -1240,6 +1241,104 @@ const ALL_AXIS_TYPES: [GamepadAxisType; 6] = [
+    }
+}
+
+/// An event that controls force-feedback rumbling of a [`Gamepad`]
```
nit doc: this needs a "vibration" somewhere. The feature has so many names, it's very easy to overlook if you are not using the right keyword.

mockersf replied:
maybe a doc alias? https://doc.rust-lang.org/rustdoc/advanced-features.html#add-aliases-for-an-item-in-documentation-search

johanhelsing commented:
> need to be careful about under/overflow on the `f32` to `u16` conversion.

Added some tests that should make sure nothing crazy happens.

I also changed how we call gilrs, so that framerate drops won't affect rumble durations.

nicopap commented on examples/input/gamepad_rumble.rs:
```diff
@@ -28,16 +28,16 @@ fn gamepad_system(
         };
 
         if button_pressed(GamepadButtonType::South) {
```
Nit: The previous comment recommended the first `if` to be the `North` condition, the second to be `East`, `South`, `West`. Well, maybe it's a bit ethnocentric, but I always read N E S W in this sequence, and it was surprising to me and more difficult to interpret the code because it didn't follow this expectation.

nicopap reviewed (changes requested):
Looks good, just need to adjust the clamping and IMO it's good to go.

nicopap reviewed (approved):
I actually didn't know that `f32 as u16` did round (therefore clamp) the value.

LGTM

B-Reif commented on crates/bevy_input/src/gamepad.rs:
````diff
@@ -1240,6 +1241,126 @@ const ALL_AXIS_TYPES: [GamepadAxisType; 6] = [
+///     }
+/// }
+/// ```
+#[doc(alias = "force feedback")]
````
```suggestion
#[doc(alias = “haptic feedback”)]
#[doc(alias = "force feedback")]
```

alice-i-cecile reviewed (changes requested):
We need to clamp in `to_gilrs_magnitude` still. Once that's resolved, this LGTM and I'll merge!

johanhelsing commented:
> We need to clamp in `to_gilrs_magnitude` still. Once that's resolved, this LGTM and I'll merge!

How so? As far as i understand it shouldn't be needed: https://github.com/rust-lang/rust/pull/71269

Or do you mean for clarity of intent?

alice-i-cecile commented:
Ah, TIL. Can we add tests for it instead then?

johanhelsing commented:
> Ah, TIL. Can we add tests for it instead then?

There already are some ;)

nicopap commented:
I think a single line comment on the `as u16` conversion in `to_gilrs_magnitude` saying "value is clamped by property of rust f32 to integer conversion" would avoid all future confusion.

alice-i-cecile commented on crates/bevy_gilrs/Cargo.toml:
```diff
@@ -11,9 +11,13 @@ keywords = ["bevy"]
 [dependencies]
 # bevy
 bevy_app = { path = "../bevy_app", version = "0.11.0-dev" }
+bevy_core = { path = "../bevy_core", version = "0.11.0-dev" }
```
```suggestion
```
